keyv/Map.h

```diff
@@ -0,0 +1,359 @@
+     * Retrieve a value for a key.
+     *
+     * @param key the key to retrieve.
+     * @return the value, or an empty string if the key is not available.
+     * @version 1.11
+     */
+    template< class V > V get( const std::string& key ) const
+        { return _get< V >( key ); }
+
+    /**
+     * Retrieve a value as a vector for a key.
+     *
+     * @param key the key to retrieve.
+     * @return the values, or an empty vector if the key is not available.
+     * @version 1.9.2
+     */
+    template< class V > std::vector< V > getVector( const std::string& key )
+        const;
+
+    /**
+     * Retrieve a value as a set for a key.
+     *
+     * @param key the key to retrieve.
+     * @return the values, or an empty set if the key is not available.
+     * @version 1.9.2
+     */
+    template< class V > std::set< V > getSet( const std::string& key ) const;
+
+    /**
+     * Asynchronously retrieve a value which to be read later.
+     *
+     * Might be implemented as a 'NOP' by backend implementations.
+     *
+     * @param key the key to retrieve.
+     * @param sizeHint the size of the value, may be ignored by implementation.
+     * @return false on error, true otherwise.
+     * @version 1.11
+     */
+    KEYV_API bool fetch( const std::string& key, size_t sizeHint=0 ) const;
+
+    /**
+     * Retrieve values from a list of keys and calls back for each found value.
+     *
+     * Depending on the backend implementation, this is more optimal than
+     * calling get() for each key.
+     *
+     * The ownership of the returned data in the callback is not transfered, so
+     * the value needs to be copied if needed.
+     *
+     * @param keys list of keys to obtain
+     * @param func callback function which is called for each found key
+     * @version 1.14
+     */
+    KEYV_API void getValues( const Strings& keys,
+                                 const ConstValueFunc& func ) const;
+
+    /**
+     * Retrieve values from a list of keys and calls back for each found value.
+     *
+     * Depending on the backend implementation, this is more optimal than
+     * calling get() for each key.
+     *
+     * The ownership of the returned data in the callback is transfered, so the
+     * data must be free'd by the caller.
+     *
+     * @param keys list of keys to obtain
+     * @param func callback function which is called for each found key
+     * @version 1.14
+     */
+    KEYV_API void takeValues( const Strings& keys,
+                                  const ValueFunc& func ) const;
+
+    /** Flush outstanding operations to the backend storage. @version 1.11 */
+    KEYV_API bool flush();
+
+    /** Enable or disable endianness conversion on reads. @version 1.11 */
+    KEYV_API void setByteswap( const bool swap );
+
+private:
+    Map( const Map& ) = delete;
+    Map( Map&& ) = delete;
+    Map& operator = ( const Map& ) = delete;
+    Map& operator = ( Map&& ) = delete;
+
+    detail::Map* const _impl;
```

Comment:
This PR is moving existing code from Lunchbox. Will do the impl, but not the rest in #2.